src/bin/sol-gpio/sol-gpio.c

```diff
@@ -0,0 +1,146 @@
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Intel Corporation nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "soletta.h"
+#include "sol-gpio.h"
+
+struct sol_gpio *gpio = NULL;
+struct sol_gpio_config config = { 0 };
+
+static void
+get_cmd(uint32_t pin)
+{
+    int ret;
+
+    SOL_SET_API_VERSION(config.api_version = SOL_GPIO_CONFIG_API_VERSION; )
+    config.dir = SOL_GPIO_DIR_IN;
+
+    gpio = sol_gpio_open(pin, &config);
+    SOL_NULL_CHECK(gpio);
+
+    ret = sol_gpio_read(gpio);
+    if (ret < 0)
+        fprintf(stderr, "%s\n", strerror(-ret));
+    else
+        fprintf(stdout, "value = %d\n", ret);
+
+    sol_gpio_close(gpio);
+}
+
+static void
+monitor_cb(void *_data, struct sol_gpio *_gpio, bool value)
+{
+    fprintf(stdout, "value = %d\n", value);
+}
+
+static void
+monitor_cmd(uint32_t pin)
+{
+    SOL_SET_API_VERSION(config.api_version = SOL_GPIO_CONFIG_API_VERSION);
+    config.dir = SOL_GPIO_DIR_IN;
+    config.in.trigger_mode = SOL_GPIO_EDGE_BOTH;
+    config.in.poll_timeout = 100;
+    config.in.cb = monitor_cb;
+
+    gpio = sol_gpio_open(pin, &config);
+    SOL_NULL_CHECK(gpio);
+
+    sol_run();
+    sol_gpio_close(gpio);
+}
+
+static void
+set_cmd(uint32_t pin, bool value)
+{
+    SOL_SET_API_VERSION(config.api_version = SOL_GPIO_CONFIG_API_VERSION);
+    config.dir = SOL_GPIO_DIR_OUT;
+    config.out.value = value;
+
+    gpio = sol_gpio_open(pin, &config);
+    SOL_NULL_CHECK(gpio);
+
+    sol_gpio_close(gpio);
+}
+
+static void
+usage(const char *program)
+{
+    fprintf(stdout,
+        "Usage:\n"
+        "   %s set [pin] [value]\n"
+        "   %s get [pin]]\n"
+        "   %s monitor [pin]\n",
+        program, program, program);
+}
+
+int
+main(int argc, char *argv[])
```

Comment:
On Mon, 07 Mar 2016 13:12:26 -0800, Gustavo Sverzut Barbieri wrote:
> > +    sol_gpio_close(gpio);
> > +}
> > +
> > +static void
> > +usage(const char *program)
> > +{
> > +    fprintf(stdout,
> > +        "Usage:\n"
> > +        "   %s set [pin] [value]\n"
> > +        "   %s get [pin]]\n"
> > +        "   %s monitor [pin]\n",
> > +        program, program, program);
> > +}
> > +
> > +int
> > +main(int argc, char *argv[])
> 
> it doesn't matter the comment, the macro is to be used.
> 
> If you don't want to run the mainloop just call `sol_quit()` from `startup()`.

That doesn't work, if you call sol_quit() before the loop is running, it
complains and does nothing else.
Also, from what I remember, the main loop implementations will check if
they have to quit and maybe do nothing of what the tool does (no idea
what that is, didn't check the code). It'd need an idler to call quit
from it.